### DISCUSSION TITLE
Training a SpanCat and NER Pipeline in spacy

### DISCUSSION THREAD

**Discussion by ChinmayR1202** [explosion/spaCy#11416] [Help: Coding & Implementations]
I want to train a spacy model with both SpanCat and NER in the pipeline. I have both NER and SpanCat as separate annotations from prodigy.

My question is that if I want to train the model, do I freeze NER first, train it on SpanCat dataset, then freeze SpanCat and train it on NER dataset? Or is there another way to do it?

**Comment by polm** [comment]
You should be able to train both of those at the same time - as long as each model has a Listener for the tok2vec component, they can share a tok2vec (or Transformer). You could also train them separately and add them into the same pipeline, like is done with two NER components in [this sample project](https://github.com/explosion/projects/tree/v3/tutorials/ner_double). I wouldn't expect it to be drastic but there may be an accuracy difference between the two approaches.

**Reply by ChinmayR1202** [reply]
Hey @polm  

Thanks for the reply! If I were to train them at the same time, does the annotation data have to be a single file? Or is there a way I can use their respective annotations and then train them at the same time? Apologies, I'm quite new to this.

**Reply by polm** [reply]
Ah, so they have separate training data? You don't have to use a single file, as with the default corpus reader you can specify a directory that contains all your training files. But each component will see all the training data, so for example if there are examples in the data for NER that should have spancat labels but don't, that could be a problem. For NER data there is a way to mark missing annotations, which would let that training data be ignored, but I'm not sure there's anything similar for spancat.

If the training data for the components is completely separate - so there's no missing labels, but the data is also pretty different - then it's unlikely sharing a tok2vec will be helpful, though you can still try it.

**Reply by polm** [reply]
Also, just for perspective, it would help if you could explain your application and give some examples of data you'd have for each component.

**Reply by ChinmayR1202** [reply]
What I want to do is use the spancat as a parser, extract key sentences from a document, and then use the NER on only those sentences. I don't want the NER to extract entities from other parts of the doc. So I thought having a spancat be a parser to clean my data and then running the NER would be one way to do it. 

I don't have any annotation data yet as I have not begun annotating, wanted to make sure how I am going to train the model and the respective annotation requirements. I'm using prodigy at the moment and learned that annotating the spancat and NER separately would be the ideal case. 

From what you said, I think it would be better to train the components separately and add them into the same pipeline right?

**Reply by polm** [reply]
Ah, OK. In that case it does sound like you should train those separately. It sounds like they should be separate pipelines, and you can pass only the flagged outputs of the spancat to the NER pipeline.

**Reply by ChinmayR1202** [reply]
Yeah, that makes sense, and guess that's the way to go. Thanks for helping me!
